src/gui/properties/piecesbar.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2016 Eugene Shalygin <eugene.shalygin@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef PIECESBAR_H
+#define PIECESBAR_H
+
+#include <QColor>
+#include <QImage>
+#include <QWidget>
+
+class QHelpEvent;
+
+namespace BitTorrent
+{
+    class TorrentHandle;
+}
+
+class PiecesBar: public QWidget
+{
+    using base = QWidget;
+    Q_OBJECT
+    Q_DISABLE_COPY(PiecesBar)
+
+public:
+    /*!
+     * @par local if true, this is downloaded pieces, avaliable otherwise
+     */
+    explicit PiecesBar(bool local, QWidget *parent = nullptr);
```

Comment:
As for `filePath.replace(QLatin1String("/.unwanted"), QString());` - it can be done unconditionally.
As for the second case - this logic belongs to subclasses actually. Bad design is more wrong than your subjective fear of virtual functions here.